utilities/boostHistHelpers.py

```diff
@@ -6,77 +6,65 @@
 
 logger = logging.child_logger(__name__)
 
-def valsAndVariances(h1, h2, allowBroadcast=True, transpose=True):
-    if not allowBroadcast and len(h1.axes) != len(h2.axes):
-        raise ValueError("Incompatible hists for math operation")
-
-    flow = True
-    for i,j in zip(h1.axes.traits, h2.axes.traits):
-        flow = i.overflow == j.overflow and i.underflow == j.underflow
-        if not flow:
-            break
-
-    if len(h1.axes) == len(h2.axes):
-        return (h1.values(flow=flow), h2.values(flow=flow), h1.variances(flow=flow), h2.variances(flow=flow))
-    else:
-        outshape = h1.view(flow=flow).shape if len(h1.shape) > len(h2.shape) else h2.view(flow=flow).shape
-        # The transpose is because numpy works right to left in broadcasting, and we've put the
-        # syst axis on the right
-        try:
-            res = [None if x is None else np.broadcast_to(x.T if transpose else x, outshape[::-1] if transpose else outshape) for x in \
-                    [h1.values(flow=flow), h2.values(flow=flow), h1.variances(flow=flow), h2.variances(flow=flow)]]
-        except ValueError as e:
-            logger.error(f"Failed to broadcast hists! h1.axes.name {h1.axes.name}, h2.axes.name {h2.axes.name}")
-            raise e
-        return [None if x is None else x.T for x in res] if transpose else res
-
-def broadcastOutHist(h1, h2):
-    if len(h1.axes) == len(h2.axes):
-        return h1 if h1.axes[-1].size >= h2.axes[-1].size else h2
-    return h1 if len(h1.axes) > len(h2.axes) else h2
+def valsAndVariances(h1, h2):
+    return h1.values(flow=True),h2.values(flow=True),h1.variances(flow=True),h2.variances(flow=True)
+
+# Broadcast h1 to match the shape of h2
+def broadcastSystHist(h1, h2):
+    if h1.ndim > h2.ndim or h1.shape == h2.shape:
+        return h1
+    
+    # Transpose because we keep syst axis last, but numpy broadcasts from the front
+    new_vals = np.broadcast_to(h1.view(flow=True).T, h2.view(flow=True).T.shape).T
+
+    return hist.Hist(*h2.axes, data=new_vals, storage=h1._storage_type())
 
 # returns h1/h2
 def divideHists(h1, h2, cutoff=1e-5, allowBroadcast=True, rel_unc=False, createNew=True):
-    # To get the broadcast shape right
-    outh = h1 if not allowBroadcast else broadcastOutHist(h1, h2)
+    if allowBroadcast:
+        h1 = broadcastSystHist(h1, h2)
+        h2 = broadcastSystHist(h2, h1)
+
+    storage = h1._storage_type() if h1._storage_type == h2._storage_type else hist.storage.Double()
+    outh = hist.Hist(*h1.axes, storage=storage) if createNew else h1
+
+    h1vals,h2vals,h1vars,h2vars = valsAndVariances(h1, h2)
+
+    # Careful not to overwrite the values of h1
+    out = outh.values(flow=True) if createNew else np.empty(outh.values(flow=True).shape)
 
-    h1vals,h2vals,h1vars,h2vars = valsAndVariances(h1, h2, allowBroadcast)
-    out = np.zeros_like(h2vals)
     # By the argument that 0/0 = 1
     out[(np.abs(h2vals) < cutoff) & (np.abs(h1vals) < cutoff)] = 1.
```

Comment:
To anticipate something I am doing, could we have a new function argument ratioValAtCutoff with default 1.0, to be used for the ratio in those bins of out where the content is smaller than cutoff? A typical use case is where the ratio represents an efficiency, in which case 0 might be more appropriate (but then the uncertainties need to be computed differently).

It can probably directly go in a new pull request, though

Reply:
Yes good point, I agree